### PR TITLE
OpenZFS - Performance regression suite for zfstest

### PR DESCRIPTION
**Pull request by don-brady**
Author: John Wren Kennedy <john.kennedy@delphix.com>
Reviewed by: Prakash Surya <prakash.surya@delphix.com>
Reviewed by: Dan Kimmel <dan.kimmel@delphix.com>
Reviewed by: Matt Ahrens <mahrens@delphix.com>
Reviewed by: Paul Dagnelie <pcd@delphix.com>
Reviewed by: Don Brady <don.brady@intel.com>
Reviewed by: Richard Elling <Richard.Elling@RichardElling.com>
Approved by: Richard Lowe <richlowe@richlowe.net>
Ported-by: Don Brady <don.brady@intel.com>

OpenZFS-issue: https://www.illumos.org/issues/6950
OpenZFS-commit: https://github.com/openzfs/openzfs/commit/dcbf3bd6
Delphix-commit: https://github.com/delphix/delphix-os/commit/978ed49

ZFS Test Suite Performance Regression Tests

This was pulled into OpenZFS via the compressed arc featureand was
separated out in zfsonlinux as a separate pull request from PR-4768.
It originally came in as QA-4903 in Delphix-OS from John Kennedy.

Porting Notes:
1. Added assertions in the setup script to make sure required tools
   (fio, mpstat, ...) are present.
2. For the config.json generation in perf.shlib used arcstats and
    other binaries instead of dtrace to query the values.
3. For the perf data collection:
   - use "zpool iostat -lpvyL" instead of the io.d dtrace script
    (currently not collecting zfs_read/write latency stats)
   - mpstat and iostat take different arguments
   - prefetch_io.sh is a placeholder that uses arcstats instead of
     dtrace
4. Build machines require fio, mdadm and sysstat packages (YMMV).

Future Work:
   - Need a way to measure zfs_read and zfs_write latencies per pool.
   - Need tools to takes two sets of output and display/graph the
     differences
   - Bring over additional regression tests from Delphix

### CODE REVIEW

File: tests/zfs-tests/include/commands.cfg.in [hunk 1]
```diff
@@ -113,8 +120,10 @@ export UNSHARE="@UNSHARE@"
 export USERADD="@USERADD@"
 export USERDEL="@USERDEL@"
 export USERMOD="@USERMOD@"
+export VMSTAT="@VMSTAT@"
 export WAIT="@WAIT@"
 export WC="@WC@"
+export ZINJECT="@ZINJECT@"
```
Comment (behlendorf) [inline tests/zfs-tests/include/commands.cfg.in]: ZINJECT doesn't need to be added here.  It's not a system tool and is already available from `zfs_commands.cfg`.  Dropping this line should fix the zpool_scrub* failures.

File: config/user-commands.m4 [hunk 2]
```diff
@@ -26,10 +26,12 @@ AC_DEFUN([ZFS_AC_CONFIG_USER_COMMANDS_COMMON], [
 	AC_PATH_TOOL(DU, du, "")
 	AC_PATH_TOOL(ECHO, echo, "")
 	AC_PATH_TOOL(EGREP, egrep, "")
+	AC_PATH_TOOL(FALSE, false, "")
 	AC_PATH_TOOL(FDISK, fdisk, "")
 	AC_PATH_TOOL(FGREP, fgrep, "")
 	AC_PATH_TOOL(FILE, file, "")
 	AC_PATH_TOOL(FIND, find, "")
+	AC_PATH_TOOL(FIO, fio, "/usr/bin/fio")
```
Comment (behlendorf) [inline config/user-commands.m4]: You shouldn't need to specify "/usr/bin/fio", this is a standard package available for most distributions.

Comment (behlendorf) [inline config/user-commands.m4]: I'd also suggest adding `Requires: fio` to the `%package test` section of the spec file, `zfs.spec.in`.  This way when the zfs-test package get's installed it will automatically pull in the fio package as a dependency.

File: config/user-commands.m4 [hunk 3]
```diff
@@ -96,12 +99,20 @@ dnl # These commands may take different command line arguments.
 dnl #
 AC_DEFUN([ZFS_AC_CONFIG_USER_COMMANDS_LINUX], [
 	AC_PATH_TOOL(BLOCKDEV, blockdev, "")
+	AC_PATH_TOOL(CHACL, chacl, "")
```
Comment (behlendorf) [inline config/user-commands.m4]: Similarly add a `Requires: acl` to pull in the acl package which provides chacl, getfacl, setfacl.

File: tests/runfiles/perf-regression.run [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+timeout = 0
+post_user = root
+post = cleanup
+outputdir = /var/tmp/test_results
+
+[tests/perf/regression]
+tests = ['sequential_writes', 'sequential_reads', 'sequential_reads_cached',
+    'sequential_reads_cached_clone', 'random_reads', 'random_writes',
+    'random_readwrite']
+post =
```
Comment (behlendorf) [inline tests/runfiles/perf-regression.run]: For the purposes of testing you could add the following line to the commit message to run the perf-regression tests instead of the usual battery.  We'll likely want to add another step to the buildbot which always, or perhaps conditionally, runs these tests depending on how long they take.

```
TEST_ZFSTESTS_RUNFILE="perf-regression.run"
```

File: config/user-commands.m4 [hunk 5]
```diff
@@ -96,12 +99,20 @@ dnl # These commands may take different command line arguments.
...
 	AC_PATH_TOOL(COMPRESS, gzip, "")
 	AC_PATH_TOOL(FORMAT, parted, "")
+	AC_PATH_TOOL(FREE, free, "")
+	AC_PATH_TOOL(GETFACL, getfacl, "")
+	AC_PATH_TOOL(IOSTAT, iostat, "")
 	AC_PATH_TOOL(LOCKFS, lsof, "")
+	AC_PATH_TOOL(LSBLK, lsblk, "")
 	AC_PATH_TOOL(MODUNLOAD, rmmod, "")
+	AC_PATH_TOOL(MPSTAT, mpstat, "")
```
Comment (behlendorf) [inline config/user-commands.m4]: And `Requires: sysstat` while we're at it.

File: tests/zfs-tests/tests/perf/regression/setup.ksh [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+. $STF_SUITE/include/libtest.shlib
+
+[[ -z  $FIO ]] || \
+[[ -z  $FREE ]] || \
+[[ -z  $IOSTAT ]] || \
+[[ -z  $LSBLK ]] || \
+[[ -z  $MPSTAT ]] || \
+[[ -z  $VMSTAT ]] && \
+    log_fail "missing required commands"
```
Comment (behlendorf) [inline tests/zfs-tests/tests/perf/regression/setup.ksh]: It would be nice to do this as 6 individual checks each with a unique error message listing the missing command.  This way when this fails you'll have an idea what command is missing.

File: tests/zfs-tests/tests/perf/fio/mkfiles.fio [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+
+#
+# Copyright (c) 2015 by Delphix. All rights reserved.
+#
+
+[global]
+filename_format=file$jobnum
```
Comment (behlendorf) [inline tests/zfs-tests/tests/perf/fio/mkfiles.fio]: Bad option <filename_format=file$jobnum> fio: job global dropped

**Comment by behlendorf** [comment]
This is awesome!  It looks like it's largely just needs some tweaking to get things running smoothly under Linux.

File: config/user-commands.m4 (hunk 3 above)
Comment (don-brady) [inline config/user-commands.m4]: Note that this PR didn't add *ACL tools, it only sorted the list per convention.

Comment (behlendorf) [inline config/user-commands.m4]: Good point, but we still really should add this dependency.  It's fine if you want to keep it in a separate patch.

File: config/user-commands.m4 (hunk 2 above)
Comment (dpquigl) [inline config/user-commands.m4]: So just to clarify you're saying we still need the AC_PATH_TOOL statement we just don't need the fallback value of /usr/bin/fio?

**Comment by dpquigl** [comment]
I stumbled across something in the packaging. Apparently perf.shlib never gets packaged in the rpm because its not listed as a distfile in that directory's makefile. I added it to the Makefile.am to make sure it gets installed. The reason this wasn't caught by the test run is because the PR doesn't actually run the regression test suite since its part of its own runfile.
